presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java

```diff
@@ -716,6 +735,16 @@ default void revokeTablePrivileges(ConnectorSession session, SchemaTableName tab
         return emptyList();
     }
 
+    /**
+     * Returns a (table name, Optional.empty()) entry if the table is redirected. See {@link #redirectTable}'s doc for details.
+     */
+    default Map<SchemaTableName, Optional<List<GrantInfo>>> listTablePrivilegesWithRedirection(ConnectorSession session, SchemaTablePrefix prefix)
```

Comment:
What's the point in returning map entries with Optional.empty?
Is _not returning_ them equally correct?

the existing code doesn't seem to use those empty entries, so if there is a future intent, it requires documenting (eg via `@apiNote`)

Reply:
`Optional.empty` means the table is redirected. Since it may be in another catalog, `MetadataManager::listTablePrivileges` should redirect it firstly and then call `ConnectorMetadata::listTablePrivileges`

Reply:
> Optional.empty means the table is redirected.

This i understand. But is it for? I didn't find any usage of those empty values, they are just filtered out . Should i search again?

Reply:
Hmm so in `MetadataManager::listTablePrivileges`:
```java
    public List<GrantInfo> listTablePrivileges(Session session, QualifiedTablePrefix prefix)
    {
        ImmutableList.Builder<GrantInfo> grantInfos = ImmutableList.builder();
        ImmutableList.Builder<QualifiedObjectName> redirectedTables = ImmutableList.builder();
        listTablePrivilegesWithRedirection(session, prefix).forEach((name, grantInfo) -> {
            if (grantInfo.isPresent()) {
                grantInfos.addAll(grantInfo.get());
            }
            else {
                redirectedTables.add(new QualifiedObjectName(
                        prefix.getCatalogName(),
                        name.getSchemaName(),
                        name.getTableName()));
            }
        });
        for (QualifiedObjectName redirectedTable : redirectedTables.build()) {
            QualifiedObjectName finalTable = redirectTable(session, redirectedTable, WarningCollector.NOOP);
            listTablePrivilegesWithRedirection(session, finalTable.asQualifiedTablePrefix()).values()
                    .forEach(tableGrantInfos -> tableGrantInfos.ifPresent(grantInfos::addAll));
        }
        return grantInfos.build();
    }
```
When the value is present, grant infos are added to `grantInfos` directly; otherwise the key is put into `redirectedTables` list. And for redirected tables, `listTablePrivilegesWithRedirection` is called again.